File: C(++)/005. Longest Palindromic Substring.c
/*
005. Longest Palindromic Substring
https://leetcode.com/problems/longest-palindromic-substring/

c 3 ms
*/

/*
Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.

Example:

Input: "babad"

Output: "bab"

Note: "aba" is also a valid answer.

Example:

Input: "cbbd"

Output: "bb"
*/

int imin(int a,int b)
{
    return a<b?a:b;
}
char* longestPalindrome(char* s) {
    char *ss=(char*)malloc(2005*sizeof(char));
    int l=strlen(s);
    for (int i=0;i<l;i++)
    {
        ss[i<<1]='\0';
        ss[i<<1|1]=s[i];
    }
    ss[l<<1]='\0';
    l=l<<1|1;
    int mr=0,le,r,id,R[2005];
    for (int i=0;i<l;i++)
    {
        if (mr>i) R[i]=imin(R[id*2-i],mr-i+1);
        else R[i]=1;
        while ((le=i-R[i])>=0&&(r=i+R[i])<l&&ss[le]==ss[r])
            R[i]++;
        if (i+R[i]-1>mr)
        {
            mr=i+R[i]-1;
            id=i;
        }
    }
    mr=0; // most right / max radius
    for (int i=0;i<l;i++)
    {
        if (R[i]>mr)
        {
            id=i;
            mr=R[i];
        }
    }
    int p=0;
    for (int i=id-R[id]+1;i<id+R[id];i++)
        if (ss[i]!='\0')
            s[p++]=ss[i];
    s[p]='\0';
    return s;
}

File: Java/005. Longest Palindromic Substring.java
/*
005. Longest Palindromic Substring
https://leetcode.com/problems/longest-palindromic-substring/

java 32 ms
*/

public class Solution {
	public int imin(int a,int b)
	{
		return a<b?a:b;
	}
	public String longestPalindrome(String s) {
		int len=s.length();
		// Java String是静态的，无法赋值或者改长度
		char []ss=new char[len*2+1];
		ss[len*2]='#'; // 首尾添加#以方便最后取出原串
		for (int i=0;i<len;i++)
		{
			ss[i*2]='#';
			ss[i*2+1]=s.charAt(i);
		}
		len=len*2+1;
		int R[]=new int[len],right=1,rpos=0,mxR=1,mxpos=0;
		// 第一个位置('#')的信息直接确定
		for (int i=1;i<len;i++)
		{
			if (i<right)
				R[i]=imin(R[rpos*2-i],right-i+1);
			else
				R[i]=1;
			int idx=i+R[i];
			while (idx<len&&(i*2-idx)>=0
				&&ss[idx]==ss[i*2-idx])
			{
				R[i]++;
				idx++;
			}
			// right以当前走到的最右为基准，即idx
			// 同时需要记录rpos用于上边的imin
			if (right<idx-1)
			{
				right=idx-1;
				rpos=i;
			}
			if (mxR<R[i])
			{
				mxR=R[i];
				mxpos=i;
			}
		}
		char ret[]=new char[mxR-1];
		// 回文串的两端一定是'#'
		right=mxpos+mxR-2;
		System.out.printf("%d %d\n",mxpos,mxR);
		for (int i=mxpos-mxR+2,j=0;i<=right;i++)
			if (ss[i]!='#')
				ret[j++]=ss[i];
		// ret.toString();返回的是地址？
		return String.valueOf(ret); // 不需要以'\0'结尾
	}
}
